bundles/org.openhab.binding.elroconnects/src/main/java/org/openhab/binding/elroconnects/internal/discovery/ElroConnectsBridgeDiscoveryService.java

```diff
@@ -0,0 +1,150 @@
+/**
+ * Copyright (c) 2010-2022 Contributors to the openHAB project
+ *
+ * See the NOTICE file(s) distributed with this work for additional
+ * information.
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ */
+package org.openhab.binding.elroconnects.internal.discovery;
+
+import static org.openhab.binding.elroconnects.internal.ElroConnectsBindingConstants.*;
+
+import java.time.Instant;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.concurrent.ScheduledFuture;
+import java.util.concurrent.TimeUnit;
+
+import org.eclipse.jdt.annotation.NonNullByDefault;
+import org.eclipse.jdt.annotation.Nullable;
+import org.openhab.binding.elroconnects.internal.ElroConnectsBindingConstants;
+import org.openhab.binding.elroconnects.internal.devices.ElroConnectsConnector;
+import org.openhab.binding.elroconnects.internal.handler.ElroConnectsAccountHandler;
+import org.openhab.core.config.discovery.AbstractDiscoveryService;
+import org.openhab.core.config.discovery.DiscoveryResultBuilder;
+import org.openhab.core.thing.Thing;
+import org.openhab.core.thing.ThingUID;
+import org.openhab.core.thing.binding.ThingHandler;
+import org.openhab.core.thing.binding.ThingHandlerService;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * {@link ElroConnectsBridgeDiscoveryService} is used to discover K1 Hubs attached to an ELRO Connects cloud account.
+ *
+ * @author Mark Herwege - Initial Contribution
+ */
+@NonNullByDefault
+public class ElroConnectsBridgeDiscoveryService extends AbstractDiscoveryService implements ThingHandlerService {
+
+    private final Logger logger = LoggerFactory.getLogger(ElroConnectsBridgeDiscoveryService.class);
+
+    private @Nullable ElroConnectsAccountHandler accountHandler;
+
+    private volatile @Nullable ScheduledFuture<?> discoveryJob;
+
+    private static final int TIMEOUT_S = 5;
+    private static final int REFRESH_INTERVAL_S = 60;
+
+    public ElroConnectsBridgeDiscoveryService() {
+        super(ElroConnectsBindingConstants.SUPPORTED_CONNECTOR_TYPES_UIDS, TIMEOUT_S);
+        logger.debug("Bridge discovery service started");
+        super.activate(null); // Makes sure the background discovery for devices is enabled
```

Comment:
This looks like a hack. I think you should call the constructor in super class taking three parameters, where last parameter is `backgroundDiscoveryEnabledByDefault` (bool). The framework will then call `startBackgroundDiscovery` when the component is activated.

Reply:
If you look at the declaration of the super constructor, it looks like it uses a default of true for this anyway, so making the call with the parameter set will not make a difference:
```
    public AbstractDiscoveryService(@Nullable Set<ThingTypeUID> supportedThingTypes, int timeout)
            throws IllegalArgumentException {
        this(supportedThingTypes, timeout, true);
    }
```

I don't know exactly where I got the description of this hack anymore, but I found it used 29 times in the repository.

Reply:
Looking at it again, I should probably call startBackgroundDiscovery directly, without the activate method. The discovery configuration parameters are irrelevant anyway. I could expose an enableBackgroundDiscovery parameter on the thing to switch it on or off (like is the default for discovery configurations using the discovery service). @jlaur WDYT?

Reply:
I implemented this idea. It turns out to be more tricky than expected. The discovery is activated before the thing handler is initialized, leading to changes in the config not being visible (without extra logic). I could implement that as well, but need to think about it.

Reply:
Modified it so background discovery is enabled/disabled from the handler, instead of discovery activation. That keeps it in sync and makes it possible to remove all initial discovery delays.